### ISSUE TITLE
Unresolved symbols in VC-WIN32 build

### ISSUE THREAD

**Issue by noahdav** [openssl/openssl#9990]
When compiling win32 in the current master branch I am unable to link  libcrypto since there are unresolved symbols.  VC-WIN64A does works as well as VC-WIN32 in the OpenSSL_1_1_1-stable branch.  According to opensslv.h it looks like this is version 3

``` C
# define OPENSSL_VERSION_MAJOR  3
# define OPENSSL_VERSION_MINOR  0
# define OPENSSL_VERSION_PATCH  0
```

Perl version v5.10.0 (ActiveState)
NASM version 2.14.02

```
perl Configure VC-WIN32 --prefix=f:\bin\openssl-1.1\x86\release --with-zlib-include=../one/oss/zlib --with-zlib-lib=zlib1.dll enable-zlib enable-zlib-dynamic

```

```                  
Configuring OpenSSL version 3.0.0-dev for target VC-WIN32                                                                                                       Using os-specific seed configuration                                                                                                                            Creating configdata.pm                                                                                                                                          Running configdata.pm                                                                                                                                           Creating makefile                                                                                                                                                                                                                                                                                                               **********************************************************************                                                                                          ***                                                                ***                                                                                          ***   OpenSSL has been successfully configured                     ***                                                                                          ***                                                                ***                                                                                          ***   If you encounter a problem while building, please open an    ***                                                                                          ***   issue on GitHub <https://github.com/openssl/openssl/issues>  ***                                                                                          ***   and include the output from the following command:           ***                                                                                          ***                                                                ***                                                                                          ***       perl configdata.pm --dump                                ***                                                                                          ***                                                                ***                                                                                          ***   (If you are new to OpenSSL, you might want to consult the    ***                                                                                          ***   'Troubleshooting' section in the INSTALL file first)         ***                                                                                          ***                                                                ***                                                                                          **********************************************************************                                                                                                                                                                           

```

```
nmake <-- this fails with the unresolved symbols
nmake install

```

```

        "d:\CxCache\Perl.Corext.5.10.0.2\x86\bin\perl.exe" util\mkdef.pl --version 3.0.0-dev --ordinals util\libcrypto.num --name libcrypto --OS windows > libcrypto.def
        IF EXIST .manifest DEL /F /Q .manifest
        IF EXIST libcrypto-3.dll DEL /F /Q libcrypto-3.dll
        link /nologo /debug /dll  /nologo /debug  /implib:libcrypto.lib /out:libcrypto-3.dll /def:libcrypto.def @C:\Users\<username>\AppData\Local\Temp\nm96BD.tmp || (DEL /Q libcrypto-3.* libcrypto.lib && EXIT 1)
   Creating library libcrypto.lib and object libcrypto.exp
libcrypto-shlib-gcm128.obj : error LNK2019: unresolved external symbol _gcm_init_clmul referenced in function _CRYPTO_gcm128_init
libcrypto-shlib-gcm128.obj : error LNK2019: unresolved external symbol _gcm_gmult_clmul referenced in function _CRYPTO_gcm128_init
libcrypto-shlib-gcm128.obj : error LNK2019: unresolved external symbol _gcm_ghash_clmul referenced in function _CRYPTO_gcm128_init
libcrypto-3.dll : fatal error LNK1120: 3 unresolved externals
NMAKE : fatal error U1077: 'link' : return code '0x1'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX64\x64\nmake.exe"' : return code '0x2'
Stop.

```


```
perl configdata.pm --dump
```

```

Command line (with current working directory = .):

    d:\CxCache\Perl.Corext.5.10.0.2\x86\bin\perl.exe .\Configure VC-WIN32 --prefix=f:\bin\openssl-1.1\x86\release --with-zlib-include=../one/oss/zlib --with-zlib-lib=zlib1.dll enable-zlib enable-zlib-dynamic

Perl information:

    d:\CxCache\Perl.Corext.5.10.0.2\x86\bin\perl.exe
    5.10.0 for MSWin32-x86-multi-thread

Enabled features:

    aria
    asm
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    engine
    err
    filenames
    fips
    gost
    idea
    legacy
    makedepend
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    pinshared
    ocb
    ocsp
    padlockeng
    pic
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    seed
    shared
    siphash
    siv
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    static-engine
    stdio
    tests
    threads
    tls
    ts
    ui-console
    uplink
    whirlpool
    zlib
    zlib-dynamic
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    ktls                    [default]   OPENSSL_NO_KTLS
    afalgeng                [not-linux] 
    asan                    [default]   OPENSSL_NO_ASAN
    buildtest-c++           [default]   
    crypto-mdebug           [default]   OPENSSL_NO_CRYPTO_MDEBUG
    crypto-mdebug-backtrace [default]   OPENSSL_NO_CRYPTO_MDEBUG_BACKTRACE
    devcryptoeng            [default]   OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128     [default]   OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                     [default]   OPENSSL_NO_EGD
    external-tests          [default]   OPENSSL_NO_EXTERNAL_TESTS
    fuzz-libfuzzer          [default]   OPENSSL_NO_FUZZ_LIBFUZZER
    fuzz-afl                [default]   OPENSSL_NO_FUZZ_AFL
    md2                     [default]   OPENSSL_NO_MD2 (skip crypto\md2)
    msan                    [default]   OPENSSL_NO_MSAN
    rc5                     [default]   OPENSSL_NO_RC5 (skip crypto\rc5)
    sctp                    [default]   OPENSSL_NO_SCTP
    ssl-trace               [default]   OPENSSL_NO_SSL_TRACE
    trace                   [default]   OPENSSL_NO_TRACE
    ubsan                   [default]   OPENSSL_NO_UBSAN
    unit-test               [default]   OPENSSL_NO_UNIT_TEST
    weak-ssl-ciphers        [default]   OPENSSL_NO_WEAK_SSL_CIPHERS
    ssl3                    [default]   OPENSSL_NO_SSL3
    ssl3-method             [default]   OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "lib",
    ARFLAGS => "/nologo",
    AS => "nasm",
    ASFLAGS => "",
    CC => "cl",
    CFLAGS => "/W3 /wd4090 /nologo /O2 /WX",
    CPP => "\$(CC) /EP /C",
    HASHBANGPERL => "/usr/bin/env perl",
    LD => "link",
    LDFLAGS => "/nologo /debug",
    MT => "mt",
    MTFLAGS => "-nologo",
    RANLIB => "CODE(0x643fd94)",
    RC => "rc",
    aroutflag => "/out:",
    asflags => "-f win32",
    asm_arch => "x86",
    asoutflag => "-o ",
    bin_cflags => "/Zi /Fdapp.pdb",
    bin_lflags => "/subsystem:console /opt:ref",
    bn_ops => "BN_LLONG",
    build_file => "makefile",
    build_scheme => [ "unified", "windows", "VC-common", "VC-WOW" ],
    cflags => "/Gs0 /GF /Gy /MD",
    coutflag => "/Fo",
    cppflags => "",
    defines => [ "ZLIB", "ZLIB_SHARED", "OPENSSL_SYS_WIN32", "WIN32_LEAN_AND_MEAN", "UNICODE", "_UNICODE", "_CRT_SECURE_NO_DEPRECATE", "_WINSOCK_DEPRECATED_NO_WARNINGS" ],
    disable => [  ],
    dso_cflags => "/Zi /Fddso.pdb",
    dso_scheme => "win32",
    enable => [  ],
    ex_libs => "ws2_32.lib gdi32.lib advapi32.lib crypt32.lib user32.lib",
    includes => [ "../one/oss/zlib" ],
    ldoutflag => "/out:",
    lflags => "",
    lib_cflags => "/Zi /Fdossl_static.pdb",
    lib_cppflags => "",
    lib_defines => [ "LIBZ=\"zlib1.dll\"", "L_ENDIAN" ],
    module_cflags => "",
    module_cxxflags => "",
    module_ldflags => "/dll",
    mtinflag => "-manifest ",
    mtoutflag => "-outputresource:",
    perl_platform => "Windows::MSVC",
    perlasm_scheme => "win32n",
    rcoutflag => "/fo",
    shared_cflag => "",
    shared_defflag => "",
    shared_defines => [  ],
    shared_ldflag => "/dll",
    shared_rcflag => "",
    shared_target => "win-shared",
    sys_id => "WIN32",
    thread_defines => [  ],
    thread_scheme => "winthreads",
    unistd => "<unistd.h>",
    uplink_arch => "common",

Recorded environment:

    AR = 
    ARFLAGS = 
    AS = 
    ASFLAGS = 
    BUILDFILE = 
    CC = 
    CFLAGS = 
    CPP = 
    CPPDEFINES = 
    CPPFLAGS = 
    CPPINCLUDES = 
    CROSS_COMPILE = 
    CXX = 
    CXXFLAGS = 
    HASHBANGPERL = 
    LD = 
    LDFLAGS = 
    LDLIBS = 
    MT = 
    MTFLAGS = 
    OPENSSL_LOCAL_CONFIG_DIR = 
    PERL = 
    RANLIB = 
    RC = 
    RCFLAGS = 
    RM = 
    WINDRES = 
    __CNF_CFLAGS = 
    __CNF_CPPDEFINES = 
    __CNF_CPPFLAGS = 
    __CNF_CPPINCLUDES = 
    __CNF_CXXFLAGS = 
    __CNF_LDFLAGS = 
    __CNF_LDLIBS = 

Makevars:

    AR              = lib
    ARFLAGS         = /nologo
    AS              = nasm
    CC              = cl
    CFLAGS          = /W3 /wd4090 /nologo /O2 /WX
    CPP             = $(CC) /EP /C
    CPPDEFINES      = 
    CPPFLAGS        = 
    CPPINCLUDES     = 
    CXXFLAGS        = 
    HASHBANGPERL    = /usr/bin/env perl
    LD              = link
    LDFLAGS         = /nologo /debug
    LDLIBS          = 
    MT              = mt
    MTFLAGS         = -nologo
    PERL            = d:\CxCache\Perl.Corext.5.10.0.2\x86\bin\perl.exe
    RC              = rc
    RCFLAGS         = 

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    makefile

build file:

    makefile

build file templates:

    Configurations\common0.tmpl
    Configurations\windows-makefile.tmpl
    Configurations\common.tmpl

```

**Comment by matbech** [comment]
I have reported and analyzed it here:
https://github.com/openssl/openssl/issues/10075
Also provides a workaround.

**Comment by slontis** [comment]
Can someone confirm if this still happens now?
@levitte does this predate the libimplementations changes?
We now have the change in bn/build.info which should fix this (maybe)
DEFINE[../../providers/libimplementations.a]=$BNDEF

**Comment by matbech** [comment]
@slontis Yes, still the same issue.

**Comment by FdaSilvaYY** [comment]
Fixed with #10222 ?

**Comment by mattcaswell** [comment]
Issue seems to have been fixed. Closing.
